Themes Showcase: Allow block-templates filter search term

jeyip opened the pull request:
### Changes proposed in this Pull Request

* The block templates filter doesn't work as expected for non `fse-eligible` stickered sites
* Props to @Addison-Stavlo for spotting the issue

**Note:**
We should consider how this will affect the Universal theme classic experience. Should universal classic users be able to search for `block-templates` right now?

### Screenshots
#### Before
![2022-02-07 10 57 57](https://user-images.githubusercontent.com/5414230/152853637-13f8d2d7-12ca-43b2-a5bc-8bba55e59bf6.gif)

#### After
![2022-02-07 10 53 56](https://user-images.githubusercontent.com/5414230/152853351-2260e8ee-6cf5-484a-9748-d75ad76310c6.gif)

### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply this PR to the local dev environment
* `yarn start`
* Visit the themes showcase for a site in `calypso.localhost:3000`
* Verify that searching for the `block-templates` filter returns the correct themes

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to https://github.com/Automattic/wp-calypso/issues/60750

jeyip committed:
Allow block-templates filter in themes showcase

Addison-Stavlo commented:
> We should consider how this will affect the Universal theme classic experience. Should universal classic users be able to search for block-templates right now?

🤔  It probably doesn't make sense for universal themes that would run in classic mode to show up under this filter. I think either we have to do some hacking or waiting:

* Hack the block-templates filter to conditionally filter out universal themes if they would run in classic mode.
* wait untill we roll out FSE to currently running universal themes.

I think the latter would be preferable if its anytime in the near-ish future, as the backend change for eligibility should fix this directly and at that point removing the conditional here in calypso would just cleanup.

jeyip committed:
Hide universal themes from block templates filter

The block templates filter is used to search for full site editing (FSE) themes. Universal themes would normally enable FSE themes. However, sites with a universal classic theme will not enable FSE, even if another universal theme is selected. Because of this, universal themes should not appear in the block templates search filter if the current site is using a universal classic theme.

jeyip reviewed (commented):
Note:

All meaningful line changes are on lines [205](https://github.com/Automattic/wp-calypso/pull/60898/files#diff-3e59937e475ddbe025a5e9ad02ba04982a7872f910c99164212098f138ac899cR205), [218](https://github.com/Automattic/wp-calypso/pull/60898/files#diff-3e59937e475ddbe025a5e9ad02ba04982a7872f910c99164212098f138ac899cR218), and [247-259](https://github.com/Automattic/wp-calypso/pull/60898/files#diff-3e59937e475ddbe025a5e9ad02ba04982a7872f910c99164212098f138ac899cR247-R259) for `client/my-sites/themes/themes-selection.jsx`

Addison-Stavlo commented on client/my-sites/themes/themes-selection.jsx:
```diff
@@ -191,69 +194,95 @@ function bindGetPremiumThemePrice( state, siteId ) {
+			// classic sites. This logic will be removed when FSE is enabled for all sites with universal themes.
+			const currentThemeId = getActiveTheme( state, siteId );
+			const currentTheme = getCanonicalTheme( state, siteId, currentThemeId );
+			const isUniversalClassicSite =
+				! blockEditorSettings?.is_fse_active && currentTheme?.template === 'blockbase';
```
I think `is_fse_eligible` might be the condition we want here? As the eligibility sticker is added in the same cases as the site option to enable universal themes for FSE. We also probably dont need the current theme check (and note as in my other comment that blockbase is not a child of blockbase and would need a separate check).

Currently this seems to detects if the user is currently using a universal theme in classic mode, but if they are not currently on a universal theme it gives no indication for whether or not activating a universal theme would result in FSE.

A bug I noticed in testing as a result:
* Take an existing non-FSE enrolled site, active a non-universal block theme like TT2.
* Notice all universal themes now show up under the block templates filter, despite that switching to them would result in classic mode.

jeyip replied:
> Currently this seems to detects if the user is currently using a universal theme in classic mode

Yes! I mistakenly thought that that was our desired outcome.

If I understand things correctly now, we'd want to hide universal themes from _any_ site on the classic experience (which, after giving it more thought, makes sense to me)

One point of confusion for me while making this change was that sites that are `is_fse_eligible: false` _can_ have the site editor enabled 😅 Once we finish the FSE roll out to everyone, which will be soon, I suppose the attribute won't matter, but I just wanted to clarify for any potential readers in the meantime.

Addison-Stavlo commented on client/my-sites/themes/themes-selection.jsx:
```diff
@@ -191,69 +194,95 @@ function bindGetPremiumThemePrice( state, siteId ) {
 
+function filterOutUniversalThemes( themes = [] ) {
+	return themes.filter( ( theme ) => {
+		return theme?.template !== 'blockbase';
```
also note blockbase does not use blockbase as a template, so the parent theme "blockbase" is not filtered out.

Addison-Stavlo reviewed (commented):
Very nice! I think this could work with some minor updates.

<s>That said, Im not sure if it makes sense to spend much time on this considering:</s>
* We currently have a "FSE" filter that will only show pure block themes, and a "universal" filter that will show universal themes. HEs are currently guiding users with these.
* It sounds like our plans are to enable FSE on universal themes early next week (cc @vindl), <s>when that happens the block-templates filter will work as expected and this change would then be unnecessary.</s> -

Update: I was wrong about this https://github.com/Automattic/wp-calypso/pull/60898#issuecomment-1036424818

Addison-Stavlo commented:
> , when that happens the block-templates filter will work as expected and this change would then be unnecessary.

Ok, I am wrong.  It would not work as expected as its currently conditioned on `isSiteUsingCoreSiteEditor` (it would only show up if actually using one of the FSE themes). So we do need an update to show the filter.

Using this PR to show the block-templates filter and conditionally filter universal on `is_fse_eligible` would set things up pretty nicely for the universal-fse rollout I think.

jeyip committed:
Consider blockbase when filtering out universal themes

jeyip committed:
Target all classic sites and not just universal classic

jeyip commented:
Responded to all comments. Ready for another review 👍

Addison-Stavlo reviewed (approved):
This is working great!  Thanks for fixing this.
